### PR TITLE
Add TriCore Architecture

### PR DESCRIPTION
**Pull request by b1llow**
Changes based on https://github.com/TriDis/capstone/tree/tricore, adaptations for new capstone versions and support for current versions of the TriCore instruction set.

The current `tc1.6.2` version of instruction set disassembly is almost complete. Other versions actually overlap with `tc1.6.2` for the most part, with only a small amount of extra content, which is still under construction.

Then it should be mentioned that basically all the `.inc` is generated using the modified `llvm-tblgen` from https://github.com/Rot127/llvm-capstone/tree/auto-sync-tblgen, except for `TriCoreGenDisassemblerTables.inc`, which modified a small part of `DecodeInstruction` and `DecodeToMCInst`

TODO:

- [x] tc1.1
- [x] tc1.2
- [x] tc1.3
- [x] tc1.3.1
- [x] tc1.6
- [x] tc1.6.1
- [x] tc1.6.2
- [x] tc1.1 tests
- [x] tc1.6.2 basic tests
- [x] others tests

### CODE REVIEW

**Comment by aquynh** [comment]
Very nice work! From which llvm is this, please?

**Comment by b1llow** [comment]
> Very nice work! From which llvm is this, please?

Maybe https://github.com/TriDis/llvm-tricore

**Comment by aquynh** [comment]
Is that unmaintained repo?

**Comment by b1llow** [comment]
Yes, and I meant to use this as a starting point only. We could probably maintain TriCore's TableGen and TriCoreDisassembler code ourselves.

**Comment by aquynh** [comment]
Can we assume that this arch is no longer developed? (Unlike Arm, x86, etc)

**Comment by b1llow** [comment]
I think it's just not as widely used.

**Comment by b1llow** [comment]
ping @XVilka

**Comment by XVilka** [comment]
@imbillow @imbillow this arch is indeed not widely used but *quite common* in the automotive and few other industries. This architecture is still being improved with ISA changes slightly from version to version, latest one being 1.6.2:
- https://wiki.qemu.org/Documentation/Platforms/TriCore

The aforementioned LLVM fork isn't complete and abandoned, and Infineon didn't indicate that they plan to implement Tricore support in the mainline LLVM or even some open fork. Thus, I suggest we focus only on the capstone and maintain it separately from the LLVM things.

**Comment by XVilka** [comment]
@imbillow note, there are slight differences for some instructions between different ISA versions, we should add the way to specify the particular ISA in the API.

**Comment by XVilka** [comment]
@Rot127 could you please also take a look, in case it will interfere with the auto-sync work?

**Comment by Rot127** [comment]
@XVilka Looks good to me.
Only the architectures which are generated with `auto-sync` will be adjusted to each other. Since you mentioned that Tricore will not be in the LLVM mainline (and is therefore not a candidate for `auto-sync`) there's no need to adjust it to the `auto-sync` design.

File: cs.c
Comment (kabeor) [inline cs.c]: plz move the last `)`to next line so we can avoid gcc warning.

**Comment by kabeor** [comment]
plz check CIFuzz.

```
Test MC/RISCV/insn-riscv64.s.cs
Test MC/TriCore/ADC_Background_Scan_1_KIT_TC275_LK.s.cs
fail CS_ARCH_TRICORE CS_MODE_TRICORE
Traceback (most recent call last):
  File "./test_corpus.py", line 134, in <module>
    test_file(fname.strip())
  File "./test_corpus.py", line 125, in test_file
    fout.write(unichr(mc_modes[(arch, mode)]))
KeyError: ('CS_ARCH_TRICORE', 'CS_MODE_TRICORE')
2023-03-25 03:08:35,059 - root - ERROR - Building fuzzers failed.
2023-03-25 03:08:35,060 - root - ERROR - Error building fuzzers for (commit: b01e5c9b2261bfad775d21b7c4b62fb5a59e229f, pr_ref: refs/pull/1973/merge).
```

**Comment by Rot127** [comment]
Could you please check my messages in Rizins Mattermost? Depending what you say regarding the `auto-sync` compatibility, I'd like to review this as well.

File: cstool/cstool.c
Comment (XVilka) [inline cstool/cstool.c]: Missing space

File: docs/TricoreInstructions.ods
Comment (XVilka) [inline docs/TricoreInstructions.ods:1]: I think this file could be removed

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by b1llow** [comment]
There should be special handling for different versions of ISA that hasn't been done yet. Currently, only TC1.6.2 is being referenced. Also, the testing may not be comprehensive enough, and I feel that there are still quite a few bugs present.

**Comment by XVilka** [comment]
@imbillow could you please also maintain a very brief TODO/status in the PR description?

**Comment by b1llow** [comment]
> @imbillow could you please also maintain a very brief TODO/status in the PR description?

done

**Comment by b1llow** [comment]
Now our disassembled mnemonic and op_str are pretty good, but inst_detail is not quite as good

**Comment by XVilka** [comment]
@kabeor could you please take a look at this to give the early feedback, since majority of the code was written, also to allow the CI jobs to run?

**Comment by XVilka** [comment]
@imbillow see also:
```
Traceback (most recent call last):
  File "./test_corpus.py", line 149, in <module>
    test_file(fname.strip())
  File "./test_corpus.py", line 64, in test_file
    "CS_MODE_TRICORE_110": CS_MODE_TRICORE_110,
NameError: global name 'CS_MODE_TRICORE_110' is not defined
2023-04-10 14:22:54,188 - root - ERROR - Building fuzzers failed.
```

**Comment by b1llow** [comment]
Because I'm not very familiar with Python 2, so I converted the `test_corpus.py` file to Python 3 syntax, resulting in the new `test_corpus3.py`. Their functionalities should be essentially the same.

However, the `cifuzz` GitHub action relies on `google/oss-fuzz` to complete the fuzz test, so it won't use `test_corpus3.py`. But I think it should still be useful, so I submitted it anyway.

**Comment by XVilka** [comment]
@kabeor what do you think about such script conversion?

**Comment by kabeor** [comment]
It's really great! Let me think about how to make the best use of it later.

**Comment by Rot127** [comment]
I hadn't time yet to add the `CAPSTONE_DIET` guards to the LLVM backends. So they are not in here as well.
@XVilka @kabeor Should this be done asap so it is part of this PR as well? Or does it have a few more days time and we maybe  update TriCore later?

**Comment by XVilka** [comment]
@Rot127 I think we can wait, yes. @imbillow what is your opinion?

**Comment by kabeor** [comment]
I think we can wait, and then I plan to release rc2.

**Comment by b1llow** [comment]
Yes, I can also wait for a few days; I'm not in a big hurry.
@Rot127 @XVilka

**Comment by Rot127** [comment]
@imbillow `CAPSTONE_DIET` is added now. The `inc` files have everything string like removed if the flag is set.

**Comment by Rot127** [comment]
Note that the `AsmWrite.inc` might not generates reproducible. Shouldn't be a problem since it should be in itself consistent. But you can also just apply the guards to be sure.

**Comment by b1llow** [comment]
@Rot127 

It looks like there are some problems with code generation without the `--printerLang=CCS` parameter.

```
llvm-tblgen: /home/aya/Source/llvm-capstone/llvm/utils/TableGen/CodeGenTarget.cpp:525: void llvm::CodeGenTarget::ComputeInstrsByEnum() const: Assertion `InstrsByEnum.size() == Insts.size() && "Missing predefined instr"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /home/aya/.local/bin/llvm-tblgen --gen-disassembler -I /home/aya/Source/llvm-project/llvm/include -I /home/aya/Source/capstone/arch/TriCore/ /home/aya/Source/capstone/arch/TriCore//TriCore.td
 #0 0x000055c6c00430b6 (/home/aya/.local/bin/llvm-tblgen+0x5ea0b6)
 #1 0x000055c6c004363b (/home/aya/.local/bin/llvm-tblgen+0x5ea63b)
 #2 0x000055c6c0041663 (/home/aya/.local/bin/llvm-tblgen+0x5e8663)
 #3 0x000055c6c0043e11 (/home/aya/.local/bin/llvm-tblgen+0x5eae11)
 #4 0x00007f36d8e51f50 (/usr/lib/libc.so.6+0x38f50)
 #5 0x00007f36d8ea08ec (/usr/lib/libc.so.6+0x878ec)
 #6 0x00007f36d8e51ea8 raise (/usr/lib/libc.so.6+0x38ea8)
 #7 0x00007f36d8e3b53d abort (/usr/lib/libc.so.6+0x2253d)
 #8 0x00007f36d8e3b45c (/usr/lib/libc.so.6+0x2245c)
 #9 0x00007f36d8e4a9f6 (/usr/lib/libc.so.6+0x319f6)
#10 0x000055c6bfc6a270 (/home/aya/.local/bin/llvm-tblgen+0x211270)
#11 0x000055c6bfa9b7c3 (/home/aya/.local/bin/llvm-tblgen+0x427c3)
#12 0x000055c6bfcbb3a8 (/home/aya/.local/bin/llvm-tblgen+0x2623a8)
#13 0x000055c6bfcbac7a (/home/aya/.local/bin/llvm-tblgen+0x261c7a)
#14 0x000055c6bfd12000 (/home/aya/.local/bin/llvm-tblgen+0x2b9000)
#15 0x000055c6bff135b3 (/home/aya/.local/bin/llvm-tblgen+0x4ba5b3)
#16 0x000055c6c007163e (/home/aya/.local/bin/llvm-tblgen+0x61863e)
#17 0x000055c6bff133f9 (/home/aya/.local/bin/llvm-tblgen+0x4ba3f9)
#18 0x00007f36d8e3c790 (/usr/lib/libc.so.6+0x23790)
#19 0x00007f36d8e3c84a __libc_start_main (/usr/lib/libc.so.6+0x2384a)
#20 0x000055c6bfa84125 (/home/aya/.local/bin/llvm-tblgen+0x2b125)
```

**Comment by b1llow** [comment]
Then there are some `unterminated #ifdef`.

**Comment by Rot127** [comment]
Weird, I can't reproduce this with ARM.

**Comment by Rot127** [comment]
Let me quickly finish up a CS commit and I take a look at it.

**Comment by Rot127** [comment]
Just fixed the missin `#endif`. But I still can't reproduce the bug from before.
I  used the `td` files from https://github.com/imbillow/capstone/tree/tricore
Did you made changes to your `td` files which are not yet pushed?

**Comment by b1llow** [comment]
Sorry about that. It's actually because I ran `llvm-tblgen` with this parameter `-I llvm-project/llvm/include`, which caused the program to `Assertion failed`. But when I use the correct parameter `-I llvm-capstone/llvm/include`, the program behaves normally.

**Comment by XVilka** [comment]
@imbillow so I assume this PR is finished, right? @kabeor could you please take a look again?

**Comment by b1llow** [comment]
> @imbillow so I assume this PR is finished, right? @kabeor could you please take a look again?

Yes. And since the `.inc` files and `.s.cs` files are almost entirely generated, they can largely be ignored when reviewing.

**Comment by kabeor** [comment]
Great，LGTM. By the way, could you please add TriCore in *Support multiple hardware architectures* in README? @imbillow 

I think it's ready to merge, any suggestion? @aquynh

**Review by kabeor — APPROVED** [review]

**Comment by Rot127** [comment]
> Yes. And since the .inc files and .s.cs files are almost entirely generated, they can largely be ignored when reviewing.

Which `inc` files were not generated?

**Comment by Rot127** [comment]
Also I suggest to add the TriCore `td` files to Capstones LLVM fork.

**Comment by b1llow** [comment]
@Rot127
> Then it should be mentioned that basically all the `.inc` is generated using the modified `llvm-tblgen` from [Rot127/llvm-capstone@`auto-sync-tblgen`](https://github.com/Rot127/llvm-capstone/tree/auto-sync-tblgen?rgh-link-date=2023-03-10T22%3A14%3A54Z), except for `TriCoreGenDisassemblerTables.inc`, which modified a small part of `DecodeInstruction` and `DecodeToMCInst`

**Comment by XVilka** [comment]
Looking forward for this to be merged.

**Comment by Rot127** [comment]
@imbillow Please open an issue after this is merged that TriCore should use the more general functions to copy the mapping data for instructions.
Just added the functions with this commit: https://github.com/capstone-engine/capstone/pull/1949/commits/518d71916b35f670050bc5f44a4fd9e58ceec3be

**Comment by XVilka** [comment]
@kabeor could this be merged?

**Comment by kabeor** [comment]
@XVilka need one more day to see if @aquynh has more thoughts.

**Comment by XVilka** [comment]
I suppose no additional thoughts 😄

**Comment by Rot127** [comment]
@imbillow Please check if TriCore is affected by https://github.com/Rot127/llvm-capstone/commit/4955bf005bba0ed1979446b60373e6d59416eda6
(I doubt it but better check this now than by an issue weeks later).

**Comment by b1llow** [comment]
> @imbillow Please check if TriCore is affected by [Rot127/llvm-capstone@4955bf0](https://github.com/Rot127/llvm-capstone/commit/4955bf005bba0ed1979446b60373e6d59416eda6) (I doubt it but better check this now than by an issue weeks later).

It seems to have little effect, for the tricore side.
Because the generated disassembler seems to use only the `DecoderTable` and `checkDecoderPredicate()` in the `TriCoreGenDisassemblerTables.inc` file to determine the `subtarget`

**Comment by XVilka** [comment]
@kabeor @aquynh sorry for being an annoyance, but it would be nice to be merged sooner than later, so the further work could be continued.

**Comment by kabeor** [comment]
some thoughts from @aquynh:

> we need to document the llvm version, so later it is easier to track
> 
> i prefer to follow what other archs are doing, that is without adding inc/ directory
>
> the code look ok, but i will need to look at all once again on anther machine where i can comment

He will comment more later :)

**Comment by XVilka** [comment]
As we noted before, there's no LLVM for Tricore, these tablegen files are "free-standing". And the autogeneration scripts that were used are part of auto-sync that is being worked separately in other PRs.

**Comment by Rot127** [comment]
> He will comment more later :)

Looking forward to it :) But here already some notes.

> i prefer to follow what other archs are doing, that is without adding inc/ directory

Other auto-sync archs will use an additional `inc` directory (in `include/capstone/`). This is necessary because we have to (or very much should) generated certain enums. But putting these enums in the `arch/<ARCH>` dir will break the architecture. Because core code (`cs.c` and the like) then needs to include headers from `arch/<ARCH>/`. Which is not wise IMHO.

> we need to document the llvm version, so later it is easier to track

This is in the making and will be added in the next week. But as XVilka said it is not really informative in case of TriCore.

**Comment by aquynh** [comment]
On Thu, Apr 20, 2023, 18:13 Anton Kochkov ***@***.***> wrote:

> As we noted before, there's no LLVM for Tricore, these tablegen files are
> "free-standing". And the autogeneration scripts that were used are part of
> auto-sync that is being worked separately in other PRs.
>

We can host a copy of your llvm in our github organization in that case.

**Comment by XVilka** [comment]
This LLVM doesn't exist, these tablegen files were written independently and separately by @imbillow

**Comment by b1llow** [comment]
Yes, I only wrote the tablegen definition for tricore ISA and did not write the LLVM part, because the main purpose was to add tricore support to Capstone.

**Comment by aquynh** [comment]
please put all the header code under include/capstone/inc/ into one file, just like the other archs.

File: arch/TriCore/TriCoreDisassembler.c
Comment (aquynh) [inline arch/TriCore/TriCoreDisassembler.c]: please follow the coding style of existing code: that is to have the open bracket { of the functions on the next line, but at the end of the line like this.

File: bindings/python/capstone/__init__.py
Comment (aquynh) [inline bindings/python/capstone/__init__.py:185]: what happens to the SH arch here?

File: cstool/cstool_tricore.c
Comment (aquynh) [inline cstool/cstool_tricore.c]: please use TABs for indentation, like existing code.

**Comment by aquynh** [comment]
please fix the coding style, then we can be ready to merge.

File: bindings/python/capstone/__init__.py
Comment (b1llow) [reply bindings/python/capstone/__init__.py:185]: Because `CS_ARCH_TRICORE=17` corresponds to C code. Then there is a missing `CS_ARCH_SH=16` in between, but actually `CS_ARCH_SH` does not have Python bindings.

**Comment by Rot127** [comment]
> please put all the header code under include/capstone/inc/ into one file, just like the other archs.

@aquynh Could you point out what the advantage of this is? It kinda breaks the idea of `auto-sync` that generated content and hand written content is clearly separated into unique files.

**Comment by b1llow** [comment]
Complete code formatting with clang-format (15.0.7) and https://github.com/torvalds/linux/blob/master/.clang-format

**Comment by kabeor** [comment]
@aquynh

File: arch/TriCore/TriCoreInstPrinter.c
Comment (aquynh) [inline arch/TriCore/TriCoreInstPrinter.c:146]: the { above does not align properly?

File: arch/TriCore/TriCoreInstPrinter.c
Comment (aquynh) [inline arch/TriCore/TriCoreInstPrinter.c:392]: this close bracket { does not align?

Comment (aquynh) [reply arch/TriCore/TriCoreInstPrinter.c:392]: ok, it seems you align case under the same column with switch.

nvm then

**Comment by aquynh** [comment]
code indentation looks better now :-)

File: bindings/python/capstone/__init__.py
Comment (aquynh) [reply bindings/python/capstone/__init__.py:185]: > > please put all the header code under include/capstone/inc/ into one file, just like the other archs.
> 
> @aquynh Could you point out what the advantage of this is? It kinda breaks the idea of `auto-sync` that generated content and hand written content is clearly separated into unique files.

i prefer to have it consistent with other archs.

is it too hard to update your tool to put all of the inc files into one?

Comment (Rot127) [reply bindings/python/capstone/__init__.py:185]: Its not really hard to change that. But it goes against the `auto-sync` architecture and is more a question of principle.

The update process is almost completely automated but internally complicated. So IMHO it is important that each update step of a file, is as isolated as much as possible from other files.

This is important because different files need different update mechanisms:

| Method | File type | Update method | Edits by hand |
|-------|------|----------------------|------------------------|
| 1 | `.inc` files | Exclusively generated by patched LLVM backends | Never/Not allowed |
| 2 | Translated LLVM C++ files | Translated by `CppTranslater` and diffing to previous file. | Rarely. Only changes which are too complicated for automation or diffing |
|3 |CS exclusive files | By hand | all |

The less dependencies these files have during the update process, the simpler `auto-sync` is.
Patching generated content (Method 1) into a by hand edited file (Method 3), adds such an dependency and more code.

> i prefer to have it consistent with other archs.

I agree that this is important, but since all LLVM based architectures (with the exception of `x86`) will be updated to use `auto-sync`, they are consistent to each other. This will add up to: `10/18 auto-sync archs` (TriCore included). And in the future even more, since `auto-sync` also eases the adding of new archs (as this PR already proved).

Comment (kabeor) [reply bindings/python/capstone/__init__.py:185]: > since all LLVM based architectures (with the exception of x86) will be updated to use auto-sync, they are consistent to each other.

I agree with this.

Comment (b1llow) [reply bindings/python/capstone/__init__.py:185]: I have an idea, maybe we can do it like this:

we insert comments like this in `<arch>.h`:

```
// generate content <section> begin
...
// generate content <section> end
```

tablgen replaces these comments with the actual content when running.

**Comment by Rot127** [comment]
This would be the way I'd implement this.
But implementing is not so much a problem.
It is the decision what is more important.

- A single header file for each architecture.

Or

- Following the auto-sync rules for file updating.

22 abr 2023 6:54:59 billow ***@***.***>:

File: bindings/python/capstone/__init__.py
Comment (Rot127) [reply bindings/python/capstone/__init__.py:185]: Something like this would be the way I'd implement this.
But implementing is not so much a problem.
It is the decision what is more important.

- A single header file for each architecture.

Or

- Following the auto-sync rules for file updating.

**Comment by b1llow** [comment]
So I need to manually **insert** the contents of the `inc` files into the `.h` file in this PR?

**Comment by Rot127** [comment]
> So I need to manually insert the contents of the inc files into the .h file in this PR?

If @aquynh says so, yes. Hadn't the time to write a patch script yet.

File: bindings/python/capstone/__init__.py
Comment (XVilka) [reply bindings/python/capstone/__init__.py:185]: I personally think we should switch to the new scheme, especially since there are plans to update other architectures, and add new ones (e.g. Xtensa, LANAI). The more architectures capstone has, the harder it will become to combine autogenerated and manual content in the same file.

Comment (XVilka) [reply bindings/python/capstone/__init__.py:185]: @aquynh so what is your decision?

Comment (kabeor) [reply bindings/python/capstone/__init__.py:185]: @XVilka Well, he insist:)

**Review by XVilka — CHANGES_REQUESTED** [review]
@imbillow could you please also update `HACK.TXT`?

File: HACK.TXT
Comment (XVilka) [inline HACK.TXT]: Seems like broken indent

**Review by XVilka — CHANGES_REQUESTED** [review]

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
One more argument against mixing autogenerated parts and manual parts inside one file - the risk of someone modifying the autogenerated part without noticing it, and `git` wouldn't be able to differentiate (e.g. [through `.gitattributes`](https://medium.com/@clarkbw/managing-generated-files-in-github-1f1989c09dfd)) between these. Separating them could help preventing this.

File: include/capstone/tricore.h
Comment (XVilka) [inline include/capstone/tricore.h]: Should we also remove the header here? I also wonder if maybe it's better to also follow the indent here

Comment (XVilka) [reply include/capstone/tricore.h]: @kabeor @aquynh @Rot127  do you agree with me? I think it's the only thing left before it could be merged, right?

**Comment by Rot127** [comment]
@XVilka Ah yes. Of cause. Removing the header is better.

26 abr 2023 4:47:29 Anton Kochkov ***@***.***>:

**Review by XVilka — APPROVED** [review]
LGTM. @aquynh @kabeor all your feedback was addressed. Your call now.

**Review by Rot127 — APPROVED** [review]

File: include/capstone/tricore.h
Comment (kabeor) [inline include/capstone/tricore.h]: please capitalize:)

**Review by kabeor — CHANGES_REQUESTED** [review]

**Comment by kabeor** [comment]
Then we can merge

File: bindings/python/capstone/tricore.py
Comment (kabeor) [inline bindings/python/capstone/tricore.py:6]: and plz capital here

File: bindings/python/capstone/tricore_const.py
Comment (kabeor) [inline bindings/python/capstone/tricore_const.py]: here

**Review by kabeor — CHANGES_REQUESTED** [review]

File: include/capstone/tricore.h
Comment (b1llow) [reply include/capstone/tricore.h]: @Rot127 Maybe we need to capitalize the target name when generating `Enum` in Capstone exports?

Comment (Rot127) [reply include/capstone/tricore.h]: Yes. Its on the list.

File: bindings/python/capstone/tricore.py
Comment (b1llow) [reply bindings/python/capstone/tricore.py:6]: > and plz capital here

image

But I noticed that PowerPC isn't capitalized either. I think there might not be such a convention?

Comment (Rot127) [reply bindings/python/capstone/tricore.py:6]: Thinking about it a second time I'm against capitalizing it. We would need to add `.upper()` pretty much everywhere in the LLVM backends when using the target name (which is done a lot). So the code gets more unreadable.
@kabeor Is there any advantage for capitalization which would make this worth it?

Comment (kabeor) [reply bindings/python/capstone/tricore.py:6]: Oh sorry, my bad. Here is good.

File: include/capstone/tricore.h
Comment (Rot127) [reply include/capstone/tricore.h]: @imbillow I pushed a commit which should capitalize the target name for all enums. Can you please test it and create a PR if something doesn't work.

Comment (b1llow) [reply include/capstone/tricore.h]: There is nothing wrong with most of them, except for these two places. I will come to PR later if I have time.

- [x] `TriCoreGenDisassemblerTables.inc#checkDecoderPredicate`
- [x] `TriCoreGenCSRegEnum.inc`
- [x] `TriCoreGenRegisterInfo.inc`

File: suite/test_corpus3.py
Comment (Rot127) [inline suite/test_corpus3.py]:
```suggestion
#!/usr/bin/env python3
```

**Comment by b1llow** [comment]
@kabeor Hi, the problem you mentioned has been fixed, please review again.

**Comment by Rot127** [comment]
@kabeor @aquynh Is it ready to be merged?

**Comment by kabeor** [comment]
@imbillow @Rot127 It's really great, can you solve the conflict so I can merge?

**Comment by XVilka** [comment]
@kabeor looks like conflict was resolved 😁

**Review by kabeor — APPROVED** [review]

**Comment by kabeor** [comment]
Thanks again for this great work! Merged.

**Comment by XVilka** [comment]
@aquynh @kabeor what do you think about tagging 5.0-rc3 with that (and release 5.0 once it's stable enough)? I wonder if it make sense to release 5.0 before auto-sync is merged, and focus 5.1 on auto-sync work?
cc @imbillow @Rot127 

Maybe addressing these two first though: https://github.com/capstone-engine/capstone/issues/2010 and https://github.com/capstone-engine/capstone/issues/2009

**Comment by aquynh** [comment]
this is what i am planning to do.

anything else do you think should be merged before rc3 tag?

**Comment by aquynh** [comment]
> @aquynh @kabeor what do you think about tagging 5.0-rc3 with that (and release 5.0 once it's stable enough)? I wonder if it make sense to release 5.0 before auto-sync is merged, and focus 5.1 on auto-sync work? cc @imbillow @Rot127
> 
> Maybe addressing these two first though: #2010 and #2009

yes, please make pull req for the above issues before rc3 tag.

my main concern now is that the Pypi package is still properly generated for all platforms.

**Comment by Rot127** [comment]
> I wonder if it make sense to release 5.0 before auto-sync is merged, and focus 5.1 on auto-sync work?

@aquynh Would agree with releasing `v5.0` before `auto-sync` and `v5.1` for updated `ARM`, `PPC` and whatever else is done at this point.

**Comment by aquynh** [comment]
> > I wonder if it make sense to release 5.0 before auto-sync is merged, and focus 5.1 on auto-sync work?
> 
> @aquynh Would agree with releasing `v5.0` before `auto-sync` and `v5.1` for updated `ARM`, `PPC` and whatever else is done at this point.

yes the auto-sync is a big change, so I want to release 5.0 before merging that.

File: bindings/python/capstone/__init__.py
Comment (peace-maker) [inline bindings/python/capstone/__init__.py:730]: Should this be `tricore.get_arch_info` instead of `riscv.get_arch_info`?

Comment (b1llow) [reply bindings/python/capstone/__init__.py:730]: You are right

Comment (b1llow) [reply bindings/python/capstone/__init__.py:730]: @peace-maker Could you please send a new PR to fix these errors?

**Pull request merged**
